vavr/src/main/java/io/vavr/collection/RedBlackTree.java

```diff
@@ -293,42 +381,145 @@
      */
     @Override
+    @SuppressWarnings({"unchecked", "rawtypes"})
     default Iterator<T> iterator() {
         if (isEmpty()) {
             return Iterator.empty();
         } else {
             final Node<T> that = (Node<T>) this;
             return new AbstractIterator<T>() {
 
-                List<Node<T>> stack = pushLeftChildren(List.empty(), that);
+                int stackCount = 0;
+                Node<T>[] stack = new Node[that.blackHeight * 2];
+
+                {
+                    pushLeftChildren(that);
+                }
 
                 @Override
                 public boolean hasNext() {
-                    return !stack.isEmpty();
+                    return stackCount != 0;
                 }
 
                 @Override
                 public T getNext() {
-                    final Tuple2<Node<T>, ? extends List<Node<T>>> result = stack.pop2();
-                    final Node<T> node = result._1;
-                    stack = node.right.isEmpty() ? result._2 : pushLeftChildren(result._2, (Node<T>) node.right);
-                    return result._1.value;
+                    // Do a pop
+                    final Node<T> node = stack[stackCount - 1];
+                    stackCount--;
+
+                    if (!node.right.isEmpty()) {
+                        pushLeftChildren((Node<T>) node.right);
+                    }
+                    return node.value;
                 }
 
-                private List<Node<T>> pushLeftChildren(List<Node<T>> initialStack, Node<T> that) {
-                    List<Node<T>> stack = initialStack;
+                private void pushLeftChildren(Node<T> that) {
                     RedBlackTree<T> tree = that;
                     while (!tree.isEmpty()) {
                         final Node<T> node = (Node<T>) tree;
-                        stack = stack.push(node);
+                        // Do a push
+                        stack[stackCount] = node;
+                        stackCount++;
                         tree = node.left;
                     }
-                    return stack;
                 }
             };
         }
     }
 
+    /**
+     * Returns an Iterator that iterates elements in the order induced by the underlying Comparator, starting from the first element
+     * that bigger or equal to the request
+     * <p>
+     * Internally an in-order traversal of the RedBlackTree is performed.
+     * <p>
+     * Example:
+     *
+     * <pre><code>
+     *       4
+     *      / \
+     *     2   6
+     *    / \ / \
+     *   1  3 5  7
+     * </code></pre>
+     *
+     * From: 4
+     * Iteration order: 4, 5, 6, 7
+     */
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    default Iterable<T> iterableFrom(T from) {
+        final Node<T> that = (Node<T>) this;
+        return new Iterable<T>() {
+            @Override
+            public java.util.Iterator<T> iterator() {
```

Comment:
What's the gain of having the custom iterator implementation here that is slightly different from the Iterator in the function above?